python/services/dataservice/dmod/dataservice/service.py

```diff
@@ -489,6 +495,96 @@ async def _async_process_query(self, message: DatasetManagementMessage) -> Datas
+        config object that includes certain details.
+
+        Parameters
+        ----------
+        request: AbstractNgenRequest
+            An AbstractNgenRequest request that needs a realization config generated, and as part of that, a forcing
+            config.
+
+        Returns
+        -------
+        Forcing
+            Forcing config object to be used in building a ngen realization config to satisfy this request.
+        """
+        # A user may work around dataset performance issues by mounting something directly from the host into
+        # containers, using an env var for the mount source (see code in scheduler.py).  We can support this by
+        # expecting a certain prefix (from_env:::) for either the file pattern or the file basename.
+        use_from_env_workaround = False
+
+        forcing_cfg_params = dict()
+
+        # Get the correct forcing dataset from associated requirement
+        # TODO: double check that this is being added when we do data checks
+        forcing_req = [r for r in request.data_requirements if r.category == DataCategory.FORCING][0]
+        forcing_dataset_name = forcing_req.fulfilled_by
+        forcing_dataset = self.get_known_datasets().get(forcing_dataset_name)
+
+        # Figure out the correct provider type from the dataset format
+        # TODO: this may not be the right way to do this to instantiate the object directly (i.e., not through JSON)
+        if forcing_dataset.data_format == DataFormat.NETCDF_FORCING_CANONICAL:
+            forcing_cfg_params['provider'] = 'NetCDF'
+        elif forcing_dataset.data_format == DataFormat.AORC_CSV:
+            forcing_cfg_params['provider'] = 'CsvPerFeature'
+
+        # TODO: (#needs_issue) introduce logic to examine forcing dataset and intelligently assess what the file
+        #  name(s)/pattern(s) should be if they aren't explicitly provided
+
+        if request.formulation_configs is not None and request.formulation_configs.forcing_file_pattern is not None:
+            forcing_cfg_params['file_pattern'] = request.formulation_configs.forcing_file_pattern
+
+        # Finally, produce the right path
+        # TODO: these come from scheduler.py; may need to centralize somehow
+        forcing_cfg_params['path'] = '/dmod/datasets/'
+        if request.formulation_configs is not None and request.formulation_configs.is_env_workaround:
+            forcing_cfg_params['path'] += 'from_env'
+        else:
+            forcing_cfg_params['path'] += '{}/{}/'.format(DataCategory.FORCING.name.lower(), forcing_dataset_name)
+
+        if request.formulation_configs is not None and request.formulation_configs.forcing_file_name is not None:
+            forcing_cfg_params['path'] += request.formulation_configs.forcing_file_name
+
+        return Forcing(**forcing_cfg_params)
+
+    def _build_ngen_realization_config_from_request(self, request: AbstractNgenRequest, job: Job) -> NgenRealization:
```

Comment:
This looks good! Is is appropriate to leave a TODO here for generating model `init_config` datasets in the future?

Reply:
maybe this makes more sense in `_derive_realization_config_from_formulations`.

Reply:
I've added a few TODO comments about making sure this gets done, including one in a different spot where I already made a similar comment about the order of things if deriving a forcing dataset is needed.